src/cache/cloud_fraction.jl

```diff
@@ -0,0 +1,233 @@
+import StaticArrays as SA
+import ClimaCore.RecursiveApply: rzero, ⊞, ⊠
+
+# TODO: write a test with scalars that are linear with z
+"""
+    Diagnose horizontal covariances based on vertical gradients
+    (i.e. taking turbulence production as the only term)
+"""
+function covariance_from_grad(coeff, mixing_length, ∇Φ, ∇Ψ)
+    return 2 * coeff * mixing_length^2 * dot(∇Φ, ∇Ψ)
+end
+
+"""
+    Compute the Smagorinsky length scale from
+    - c_smag coefficient
+    - N_eff - buoyancy frequency = sqrt(max(ᶜlinear_buoygrad, 0))
+    - dz - vertical grid scale
+    - Pr - Prandtl number
+    - ϵ_st - strain rate norm
+"""
+function smagorinsky_length_scale(c_smag, N_eff, dz, Pr, ϵ_st)
+    FT = eltype(c_smag)
+    #return N_eff > FT(0) && N_eff < sqrt(2 * Pr * ϵ_st) ?
+    return N_eff > FT(0) ?
+           c_smag * dz * max(0, (1 - N_eff^2 / Pr / 2 / ϵ_st))^(1 / 4) :
```

Comment:
```suggestion
           c_smag * dz * max(0, (1 - N_eff^2 / Pr / 2 / max(ϵ_st, eps(FT)))^(1 / 4) :
```
Just to be safe. You don't have an issue right now as you are calling it during the diagnostics, when ϵ_st is probably never zero. But if we call it at the first time step (which we may do in the end), it will cause problems as we start with zero velocities.